internal/lsp/file_handler.go

```diff
@@ -60,6 +95,13 @@ func (fh *versionedFileHandler) Version() int {
 	return fh.v
 }
 
-func FileHandlerFromPath(path string) FileHandler {
-	return FileHandler(filesystem.URIFromPath(path))
+func FileHandlerFromPath(path string) *fileHandler {
+	return &fileHandler{uri: filesystem.URIFromPath(path)}
+}
+
+func FileHandlerFromDirPath(dirPath string) *fileHandler {
+	// Dir URIs are usually without trailing separator in LSP
+	dirPath = strings.TrimSuffix(dirPath, string(os.PathSeparator))
```

Comment:
I think you could just simplify this as `filepath.Clean` unless you explicitly don't want it to resolve `.` and `..` references?

```suggestion
	dirPath = filepath.Clean(dirPath)
```